File: iOS/Pulsar/Sources/Pulsar/Audio/AudioTypes.swift
import Foundation

@objc public class PatternPoint: NSObject, Codable {
  let time: Double
  let value: Float
  @objc public init(time: Double, value: Float) {
    self.time = time
    self.value = value
  }
}
 
@objc public class DiscretePoint: NSObject, Codable {
  let time: Double
  let amplitude: Float
  let frequency: Float
  @objc public init(time: Double, amplitude: Float, frequency: Float) {
    self.time = time
    self.amplitude = amplitude
    self.frequency = frequency
  }
}

@objc public class ContinuesPattern: NSObject, Codable {
  let amplitude: [PatternPoint]
  let frequency: [PatternPoint]
  @objc public init(amplitude: [PatternPoint], frequency: [PatternPoint]) {
    self.amplitude = amplitude
    self.frequency = frequency
  }
}

@objc public class PatternData: NSObject, Codable {
  let continuesPattern: ContinuesPattern
  let discretePattern: [DiscretePoint]
  @objc public init(continuesPattern: ContinuesPattern, discretePattern: [DiscretePoint]) {
    self.continuesPattern = continuesPattern
    self.discretePattern = discretePattern
  }
  public init(line: [[[Double]]], bar: [[Double]]) {
    self.continuesPattern = ContinuesPattern(
      amplitude: line[0].map { PatternPoint(time: Double($0[0]), value: Float($0[1])) },
      frequency: line[1].map { PatternPoint(time: Double($0[0]), value: Float($0[1])) }
    )
    self.discretePattern = bar.map { DiscretePoint(time: Double($0[0]), amplitude: Float($0[1]), frequency: Float($0[2])) }
  }
}

public enum WaveformType: String {
	case sine = "sine"
	case square = "square"
	case triangle = "triangle"
	case sawtooth = "sawtooth"
}

struct OscillatorConfig {
	let frequency: (initial: Double, final: Double, decayTime: Double)
	let envelope: (attack: Double, decay: Double, sustainLevel: Double, sustainDuration: Double, release: Double)
	let waveform: WaveformType
}

struct DiscreteAudioConfig {
	let oscillator: OscillatorConfig
	let timestamp: Double
	let volume: Float
}

struct ContinuousAudioConfig {
	let type: String
	let data: (amplitude: [PatternPoint], frequency: [PatternPoint])
}

struct AudioPatternConfig {
	let discreteData: [DiscreteAudioConfig]
	let continuousData: [ContinuousAudioConfig]
}

File: Android/Pulsar/src/main/java/com/swmansion/pulsar/audio/AudioTypes.kt
package com.swmansion.pulsar.audio

data class PatternPoint(
    val time: Float,
    val value: Float
)

data class DiscretePoint(
    val time: Float,
    val amplitude: Float,
    val frequency: Float
)

data class ContinuesPattern(
    val amplitude: List<PatternPoint>,
    val frequency: List<PatternPoint>
)

data class PatternData(
    val continuesPattern: ContinuesPattern,
    val discretePattern: List<DiscretePoint>
) {
    constructor(line: List<List<List<Double>>>, bar: List<List<Double>>) : this(
        continuesPattern = ContinuesPattern(
            amplitude = line[0].map { PatternPoint(time = it[0].toFloat(), value = it[1].toFloat()) },
            frequency = line[1].map { PatternPoint(time = it[0].toFloat(), value = it[1].toFloat()) }
        ),
        discretePattern = bar.map { DiscretePoint(time = it[0].toFloat(), amplitude = it[1].toFloat(), frequency = it[2].toFloat()) }
    )
}

enum class WaveformType(val value: String) {
    SINE("sine"),
    SQUARE("square"),
    TRIANGLE("triangle"),
    SAWTOOTH("sawtooth")
}

data class OscillatorConfig(
    val frequency: FrequencyConfig,
    val envelope: EnvelopeConfig,
    val waveform: WaveformType
)

data class FrequencyConfig(
    val initial: Double,
    val final: Double,
    val decayTime: Double
)

data class EnvelopeConfig(
    val attack: Double,
    val decay: Double,
    val sustainLevel: Double,
    val sustainDuration: Double,
    val release: Double
)

data class DiscreteAudioConfig(
    val oscillator: OscillatorConfig,
    val timestamp: Double,
    val volume: Float
)

data class ContinuousAudioConfig(
    val type: String,
    val data: AudioDataConfig
)

data class AudioDataConfig(
    val amplitude: List<PatternPoint>,
    val frequency: List<PatternPoint>
)

data class AudioPatternConfig(
    val discreteData: List<DiscreteAudioConfig>,
    val continuousData: List<ContinuousAudioConfig>
)
